subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -207,35 +180,23 @@ public boolean isDependencyToAll(String className) {
      * further dependents, while resources are just data accumulated along the way. Recurses for classes that
      * are "publicly accessbile", i.e. classes that are not just used privately in a class.
      */
-    private void processDependentClasses(Set<String> visitedClasses,
-                                         Set<String> privateResultClasses,
-                                         Set<String> accessibleResultClasses,
-                                         Set<GeneratedResource> resultResources,
-                                         Iterable<String> privateDependentClasses,
-                                         Iterable<String> accessibleDependentClasses) {
-
-        for (String privateDependentClass : privateDependentClasses) {
-            if (!visitedClasses.add(privateDependentClass)) {
-                continue;
-            }
-            privateResultClasses.add(privateDependentClass);
-            DependentsSet currentDependents = getDependents(privateDependentClass);
-            if (!currentDependents.isDependencyToAll()) {
-                resultResources.addAll(currentDependents.getDependentResources());
-            }
+    private String processDependentClasses(Set<String> visitedClasses,
+                                           Set<String> privateResultClasses,
+                                           Set<String> accessibleResultClasses,
+                                           Set<GeneratedResource> resultResources) {
+        String fullRecompileReason = processTransitiveDependentClasses(visitedClasses, accessibleResultClasses, privateResultClasses, resultResources);
```

Comment:
Previously, we did the non-transitives first, which meant that they were in `visitedClasses` and would prevent them from being processed by the transitive analysis. Since classes can be both private and accessible dependents, it is important to check accessible first.